File: 01-dsa-practices/03-session_2026-05-03/_code/1_primes/PrimeNumbersSieve.java
import java.util.ArrayList;
import java.util.List;

public class PrimeNumbersSieve {

    /**
     * Generate all prime numbers up to n using Sieve of Eratosthenes.
     *
     * @param n Upper bound (inclusive)
     * @return List of prime numbers ≤ n
     */
    public static List<Integer> sieveOfEratosthenes(int n) {
        List<Integer> primes = new ArrayList<>();

        if (n < 2) {
            return primes;
        }

        boolean[] isPrime = new boolean[n + 1];

        for (int i = 0; i <= n; i++) {
            isPrime[i] = true;
        }

        isPrime[0] = false;
        isPrime[1] = false;

        for (int i = 2; i * i <= n; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= n; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        for (int i = 2; i <= n; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }

        return primes;
    }

    public static void main(String[] args) {
        int n = 30;
        List<Integer> primes = sieveOfEratosthenes(n);
        System.out.println("Primes up to " + n + ": " + primes);
    }
}

File: 01-dsa-practices/03-session_2026-05-03/_code/1_primes/prime_numbers_sieve.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * Generate all prime numbers up to n using Sieve of Eratosthenes.
 */
vector<int> sieveOfEratosthenes(int n) {
    vector<int> primes;

    if (n < 2) {
        return primes;
    }

    vector<bool> isPrime(n + 1, true);
    isPrime[0] = isPrime[1] = false;

    for (int i = 2; i * i <= n; i++) {
        if (isPrime[i]) {
            for (int j = i * i; j <= n; j += i) {
                isPrime[j] = false;
            }
        }
    }

    for (int i = 2; i <= n; i++) {
        if (isPrime[i]) {
            primes.push_back(i);
        }
    }

    return primes;
}

int main() {
    int n = 30;
    vector<int> primes = sieveOfEratosthenes(n);

    cout << "Primes up to " << n << ": ";
    for (int p : primes) {
        cout << p << " ";
    }
    cout << endl;

    return 0;
}
